File: leetcode/number-of-islands.java
//https://leetcode.com/problems/number-of-islands/

class Solution {
    public int numIslands(char[][] grid) {
		// Can be optimized for constant space by manipulating grid directly
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        int counter = 0;

        for(int i = 0; i < grid.length; i++){
            for(int j = 0; j<grid[0].length; j++){
                if(grid[i][j] == '1' && !visited[i][j]) {
                    recurse(i, j, visited, grid);
                    counter++;
                }
            }
        }	

        return counter;
    }	

    private void recurse(int row, int col, boolean[][] visited, char[][] grid) {
        if(	row >= 0 && row < grid.length && col >= 0 && col < grid[0].length && !visited[row][col]){
            visited[row][col] = true;
            if(grid[row][col] == '1'){
                recurse(row + 1, col, visited, grid);
                recurse(row - 1, col, visited, grid);
                recurse(row, col + 1, visited, grid);
                recurse(row, col - 1, visited, grid);
            }
        }
    }
}

File: leetcode/number_of_islands.py
# https://leetcode.com/problems/number-of-islands/description/


class Solution:
    def numIslands(self, grid: list[list[str]]) -> int:
        islands: int = 0

        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == "1":
                    islands += 1
                    self.flood(row, col, grid)

        return islands

    def flood(self, row: int, col: int, grid: list[list[str]]) -> None:
        if (
            row >= 0
            and col >= 0
            and row < len(grid)
            and col < len(grid[0])
            and grid[row][col] == "1"
        ):
            grid[row][col] = "0"
            self.flood(row + 1, col, grid)
            self.flood(row - 1, col, grid)
            self.flood(row, col + 1, grid)
            self.flood(row, col - 1, grid)
